src/kmeans.jl

```diff
@@ -64,29 +62,30 @@ end
 
 # core k-means skeleton
 function _kmeans!(
-    x::Matrix{T},                   # in: sample matrix (d x n)
-    w::Union{Nothing, Vector{T}},      # in: sample weights (n)
-    centers::Matrix{T},             # in/out: matrix of centers (d x k)
-    assignments::Vector{Int},       # out: vector of assignments (n)
-    costs::Vector{T},               # out: costs of the resultant assignments (n)
-    counts::Vector{Int},            # out: the number of samples assigned to each cluster (k)
-    cweights::Vector{Float64},      # out: the weights of each cluster
-    maxiter::Int,                   # in: maximum number of iterations
-    tol::Real,                      # in: tolerance of change at convergence
-    displevel::Int,                 # in: the level of display
-    distance::SemiMetric) where T<:AbstractFloat             # in: function to calculate the distance with
+    X::AbstractMatrix{<:Real},        # in: sample matrix (d x n)
+    w::Union{Nothing, AbstractVector{<:Real}}, # in: sample weights (n)
+    centers::AbstractMatrix{<:Real},  # in/out: matrix of centers (d x k)
+    assignments::Vector{Int},     # out: vector of assignments (n)
+    costs::Vector{Float64},       # out: costs of the resultant assignments (n)
+    counts::Vector{Int},          # out: # samples assigned to each cluster (k)
+    cweights::Vector{Float64},    # out: weights of each cluster
+    maxiter::Integer,             # in: maximum number of iterations
+    tol::Real,                    # in: tolerance of change at convergence
+    displevel::Int,               # in: the level of display
+    distance::SemiMetric          # in: function to calculate the distance with
+    )
 
     # initialize
-
     k = size(centers, 2)
     to_update = Vector{Bool}(undef, k) # indicators of whether a center needs to be updated
-    unused = Int[]
+    unused = Vector{Int}()
     num_affected::Int = k # number of centers, to which the distances need to be recomputed
 
-    dmat = pairwise(distance, centers, x)
-    dmat = convert(Array{T}, dmat) #Can be removed if one day Distance.result_type(SqEuclidean(), T, T) == T
```

Comment:
Are you sure this is the case now?

Reply:
`pairwise` takes an abstractmatrix, modifies it in place and returns it (https://github.com/JuliaStats/Distances.jl/blob/c4431c025790d11a85171e383e56b48f8d3d78c5/src/generic.jl#L84_L95 so I don't think there needs to be a conversion here. However `update_assignments` needs to take an `AbstractMatrix`, I've changed that.

Reply:
No, that's `pairwise!`. `pairwise` allocates a ` Matrix{result_type(metric, a, b)}`. Unless you really know what you are doing, better keep this, anyway it will be a no-op if the type is already correct.

Reply:
ok!

Reply:
apologies for being a PITA here but what's the proper type? surely not elements of `X` or `centers`, it's unrelated and `dmat` is only assigned via `pairwise` and only used in the function `update_assignments!` where the only thing that matters is that its elements can be compared between themselves so anything `<:Real` is definitely fine.
So given we know `pairwise` returns a `Matrix` of a type `<:Real` and as long as `update_assignments!` takes exactly that, I'm not sure I see a point where this convert will be useful?

Reply:
TBH I don't really know, one would need to check the history and see whether the function is type stable under various combinations of types for `X` and `centers`. As I noted in another comment, if `eltype(X) === Float32`, computations can be faster if other matrices are also in `Float32` rather than `Float64`, so that may be a reason. Another one is that `dmat` is reused below and to avoid type unstability it must not change its type.

Anyway, a general rule is that when you don't understand why a line is there, either keep it as-is, or investigate until you understand why it's been added.